bika/lims/browser/auditlog.py

```diff
@@ -0,0 +1,309 @@
+from bika.lims.browser.bika_listing import BikaListingView
+from bika.lims.interfaces import IAuditable
+from bika.lims.utils import t
+from plone.memoize import view
+from Products.CMFPlone.i18nl10n import ulocalized_time
+from Products.Five.browser import BrowserView
+from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
+
+
+class HasAuditLog(BrowserView):
+    """Helper View to check if the context provides the IAuditable interface
+    """
+    def __call__(self):
+        return IAuditable.providedBy(self.context)
+
+
+class AuditLogView(BikaListingView):
+    """Audit View
+    """
+    template = ViewPageTemplateFile("templates/audit.pt")
+    diff_template = ViewPageTemplateFile("templates/audit_diff.pt")
+
+    def __init__(self, context, request):
+        super(AuditLogView, self).__init__(context, request)
+
+        self.show_select_column = False
+        self.show_search = False
+        self.pagesize = 5
+
+        self.icon = "{}/{}/{}".format(
+            self.portal_url,
+            "++resource++bika.lims.images",
+            "%s_big.png" % context.portal_type.lower())
+
+        self.title = "{} Log".format(api.get_title(context))
+
+        self.columns = collections.OrderedDict((
+            ("version", {
+                "title": _("Version"), "sortable": False}),
+            ("time", {
+                "title": _("Time"), "sortable": False}),
+            ("modified", {
+                "title": _("Date Modified"), "sortable": False}),
+            ("actor", {
+                "title": _("Actor"), "sortable": False}),
+            ("remote_address", {
+                "title": _("Remote IP"), "sortable": False}),
+            ("action", {
+                "title": _("Action"), "sortable": False}),
+            ("review_state", {
+                "title": _("Workflow State"), "sortable": False}),
+            ("comment", {
+                "title": _("Comment"), "sortable": False}),
+            ("diff", {
+                "title": _("Changes"), "sortable": False}),
+        ))
+
+        self.review_states = [
+            {
+                "id": "default",
+                "title": "All",
+                "contentFilter": {},
+                "columns": self.columns.keys(),
+            }
+        ]
+
+    def update(self):
+        """Update hook
+        """
+        super(AuditLogView, self).update()
+
+    def make_empty_item(self, **kw):
+        """Create a new empty item
+        """
+        item = {
+            "uid": None,
+            "before": {},
+            "after": {},
+            "replace": {},
+            "allow_edit": [],
+            "disabled": False,
+            "state_class": "state-active",
+        }
+        item.update(**kw)
+        return item
+
+    @view.memoize
+    def get_snapshots(self):
+        """Get all snapshots from the storage
+
+        :returns: List of snapshots
+        """
+        from bika.lims.subscribers.auditlog import get_storage
```

Comment:
Why to not move this import at the top of the class?

Reply:
Yes, I named the method before the same as the import. However, I left it there because I don't like to import from `subscribers` and this reminds me to refactor that block